### ISSUE TITLE
4.9.12.2. Mismatch between cell header algorithm, TD HEADERs, and accessibility APIs

### ISSUE THREAD

**Issue by dd8** [w3c/html#583]
The algorithm in "4.9.12.2. Forming relationships between data cells and header cells" assumes a single header list for cells (as did HTML 4)

Modern accessibility APIs expose a pair of properties for cell headers: RowHeaders and ColumnHeaders. This allows screen readers to voice column headers when moving along a row, voice row headers when moving up or down a column, and provide separate commands to announce row or column header.

This surfaces in ARIA as the columnheader and rowheader roles

This presents a problem for browser implementors trying to match up the single header list specified by the algorithm and the RowHeaders and ColumnHeaders lists exposed by accessibility APIs.

The biggest problem seems to be the TD HEADERS attribute since there's no way to specify whether the attribute references a row header or column header (NVDA has some clever heuristics, Safari just assumes they're column headers)

All of this lead to interop issues between different browsers / ATs.

This post has a lot of detail on current browser / AT behaviour along with links to underlying browser / AT source code:
http://www.powermapper.com/blog/table-headers-and-scope/

**Comment by stevefaulkner** [comment]
@cookiecrook , @asurkov, @cyns ping

**Comment by cookiecrook** [comment]
@dd8 Are there demonstrable user problems with the implementation differences across UAs? Is this causing a challenge for validation?

**Comment by dd8** [comment]
Here is the TD HEADERS example in WCAG Sufficient Technique H43:
https://www.w3.org/TR/WCAG20-TECHS/H43.html

And here's how it behaves in different screen readers - results range from fairly easy to understand to very hard to understand:
http://www.powermapper.com/tests/screen-readers/tables/table-td-headers/

A good test for user impact is trying to answer the question: "what percentage of the total mark does the final exam count for"

Arguably the H43 table could be restructured to make it easier to understand. This isn't always possible - the Periodic Table of the Elements can't be restructured and it's difficult to make it work whilst preserving the tabular relationships. School chemistry test questions often show a periodic table and ask questions which depend on the tabular relationships:

"Which of the following elements has the most protons: Palladium, Vanadium, ..." 
"Name a radioactive element with similar chemical properties to Neon"
https://en.wikipedia.org/wiki/Periodic_table#/media/File:14LaAc_periodic_table_IIb.jpg

**Comment by asurkov** [comment]
Is the point that the HTML spec would benefit if [1] alg would deal row headers and column headers separately? If so, then I agree it'd be definitely good for accessibility.

I'm not sure though which document (HTML or HTML-AAM) should be more suitable for this.

[1] https://html.spec.whatwg.org/multipage/tables.html#header-and-data-cell-semantics

**Comment by dd8** [comment]
Yes, exactly so. 

At the moment the single headers algorithm can't be used by Accessibility APIs with separate row headers and column headers properties which means it can't be used by: 
- IAccessible2 IAccessibleTableCell (Firefox and Linux)
- UIA ITableItemProvider (Windows)
- NSAccessibility (Mac/iOS)

I'm not sure about the situation with legacy MSAA or the Android accessibility API.

**Comment by asurkov** [comment]
I cannot speak for HTML spec, I'm not sure for example, what that headers algorithm is used for, and would it harm anyone if it was split into two: row and column headers. Having said that I think it would be definitely helpful for the accessibility, and would help to synchronize implementations between browsers.

**Comment by stevefaulkner** [comment]
@dd8 can you propose changes to the specification text? If so a PR is welcome

**Comment by dd8** [comment]
Ok, I'll try to draft some changes. I'll need to investigate what MSAA and the Android API does with table headers first.

**Comment by stevefaulkner** [comment]
@dd8 any progress?

**Comment by chaals** [comment]
I'm not sure what tools use the headings algorithms, apart from browsers reporting them for screenreaders. 

That question applies also to #819 which is sort of a subset of this. Knowing which tools were using this - data processors? - would help understand what the risk of changing it is.

**Comment by chaals** [comment]
Unassigned @stevefaulkner for now because he is stepping down as an editor. Feel free to pick this up and make it a Pull Request.

**Comment by siusin** [comment]
We're closing this issue on the W3C HTML specification because the W3C and WHATWG are now [working together on HTML](https://www.w3.org/blog/2019/05/w3c-and-whatwg-to-work-together-to-advance-the-open-web-platform/), and all issues are being discussed on the [WHATWG repository](https://github.com/whatwg/html).

If you have questions about this, please open an issue on the [W3C HTML WG repository](https://github.com/w3c/htmlwg/issues/new) or send an email to public-html@w3.org.
